src/board.rs

```diff
@@ -149,40 +149,37 @@ impl Board {
     }
 
     pub fn neighbouring_squares(&self, position: Coordinate) -> Vec<(Coordinate, Square)> {
-        let mut neighbours = Vec::new();
-        for delta in Direction::iter() {
-            let neighbour_coordinate = position.add(delta);
-            if let Ok(square) = self.get(neighbour_coordinate) {
-                neighbours.push((neighbour_coordinate, square));
-            };
-        }
-        neighbours
+        Direction::iter()
+            .filter_map(|delta| {
+                let neighbour_coordinate = position.add(delta);
+                if let Ok(square) = self.get(neighbour_coordinate) {
+                    Some((neighbour_coordinate, square))
+                } else {
+                    None
+                }
+            })
+            .collect()
     }
 
     // TODO: return iterator or rename since it doesn't matter that this is depth first when we return a HashSet
     fn depth_first_search(&self, position: Coordinate) -> HashSet<Coordinate> {
-        let mut set = HashSet::new();
-
-        let player = if let Ok(Square::Occupied(player, _)) = self.get(position) {
-            player
-        } else {
-            return set;
-        };
-        let mut stack = vec![position]; // TODO: consider more efficient stack type
-
-        while let Some(current) = stack.pop() {
-            set.insert(current);
-            for (position, square) in self.neighbouring_squares(current) {
-                // Put the neighbour in the set if it is occupied by the current player
-                if let Square::Occupied(neighbours_player, _) = square {
-                    if !set.contains(&position) && player == neighbours_player {
-                        stack.push(position);
+        let mut visited = HashSet::new();
+
+        fn dfs(b: &Board, position: Coordinate, visited: &mut HashSet<Coordinate>) {
```

Comment:
I want to keep the recursive step hidden from the outside to reduce clutter, but it feels kinda weird declaring a `fn` inside a `fn`. Do you think this is a reasonable pattern?

Reply:
I can poke at this tonight — yes it's a reasonable pattern, but I normally see it implemented with closures rather than nested fns. So something like:

```rust
let dfs = |board, position, visited| {
...
}
```

_(You can usually omit a lot of the type signatures which helps clean things up, as they're inferred by usage.)_

I can rework it tonight though and provide you with a working example

Reply:
In saying that, the _recursive_ closure might be more challenging, in which case the nested fn is (maybe) ok

Reply:
But I think you can just impl `dfs` on `Board` — it will only be visible to the outside if it were a `pub fn`, so I don't think it clutters

Reply:
You can also write multiple impl blocks. So you could do

```rust
impl Board {
  pub fn something(&self) {
     ...
  }
}

impl Board {
  fn internal(&self) {
     ...
  }
}
```

Which helps when wading through impl blocks

Reply:
Upon reflection it's perfectly fine how you've written it — I quite like it